Proposal for a new UserInterface

ezimuel opened the pull request:
This PR proposes a new [UserInterface](https://github.com/ezimuel/zend-expressive-authentication/blob/feature/new-user-interface/src/UserInterface.php) and a [DefaultUser](https://github.com/ezimuel/zend-expressive-authentication/blob/feature/new-user-interface/src/DefaultUser.php) implementation.

**UPDATE** see the update commits and comment [here](https://github.com/zendframework/zend-expressive-authentication/pull/28#issuecomment-390711855).

This PR fixes #24 and it is a replacement of #25.

**NOTE:** this PR contains BC break. It also requires to update the `zend-expressive-authentication` adapters like [zend-expressive-authentication-oauth2](https://github.com/zendframework/zend-expressive-authentication-oauth2),  [zend-expressive-authentication-session](https://github.com/zendframework/zend-expressive-authentication-session), [zend-expressive-authentication-zendauthentication](https://github.com/zendframework/zend-expressive-authentication-zendauthentication), [zend-expressive-authentication-basic](https://github.com/zendframework/zend-expressive-authentication-basic).

ezimuel committed:
New UserInterface

ezimuel committed:
Fixed CS issue

ezimuel committed:
Added DefaultUser test

geerteltink commented on src/UserInterface.php:
```diff
@@ -11,15 +11,25 @@
      * @return string[]
      */
-    public function getUserRoles() : array;
+    public function getRoles() : array;
```
👍

geerteltink reviewed (commented):
This is a step in the right direction, but it doesn't fix #24 and it's not really a replacement for #25. This is actually what I had in an early stage in #25. Basically the problem to store custom data in the session still exists. Only the identity and the user roles are stored. There is no method in the user interface the session persistence classes can rely on to (re)store additional data.

The idea in #24 and #25 is to store additional data to the identity and roles. Let's say I have a User object and want to persist it in a session with this data:
```
[
    'identity' => 'me@example.com',
    'name' => 'xtreamwayz',
    'roles' => [],
    'claims' => [], 
]
```
I don't see how zend-expressive-authentication-session knows how to (re)store the additional data with the changes in this PR.

I like the changes in this PR and I think it is a step in the right direction, but I think it doesn't go far enough.

weierophinney commented on docs/book/v1/intro.md:
```diff
@@ -42,6 +55,33 @@ if a user has been authenticated or not, e.g. it can be used to verify the
+## Default User class
+
+We provided a default user class, implemented by `Zend\Authentication\DefaultUser`.
+This class is a basic implementation of `UserInterface`. This default class
```
Replace the first sentence with:

```markdown
We provide a default implementation of `UserInterface` via the class `Zend\Expressive\Authentication\DefaultUser`.
```

weierophinney commented on docs/book/v1/intro.md:
```diff
@@ -23,17 +23,30 @@ interface UserInterface
+    /**
+     * Set the user identity
+     */
+    public function setIdentity(string $identity) : void
```
I'm a little skeptical of adding setter methods to the `UserInterface`. The reason is because this could potentially lead to cases where middleware resets the identify, roles, etc. — which is fine for altering the user instance passed _deeper_ into the request layer, but could have ramifications for _parent_ layers as well. I think we should prefer immutability here.

weierophinney commented on src/ConfigProvider.php:
```diff
@@ -58,11 +59,15 @@ public function getDependencies() : array
                 UserRepository\PdoDatabase::class => UserRepository\PdoDatabaseFactory::class
+            ],
+            'invokables' => [
+                DefaultUser::class => DefaultUser::class
```
What you and I had discussed earlier was that the user "service" would map to a _factory class_ that the authentication middleware would then compose; by doing that, we both de-couple how the instance is produced as well as the type produced.

I wonder if that might answer some of the questions from @xtreamwayz , as individual adapters could provide their own user implementations **and** the factories for producing them.

geerteltink replied:
It's a step closer to a solution. However, it's not only about a User object. It's also about storing in and retrieving from a session.

weierophinney commented on src/DefaultUser.php:
```diff
@@ -0,0 +1,37 @@
+
+namespace Zend\Expressive\Authentication;
+
+class DefaultUser implements UserInterface
```
I would make this class `final`, and mark the properties as `private`. The interface is pretty slim, and it's less likely that a developer will extend this class than provide a clean implementation.

ezimuel commented:
@xtreamwayz I think the need to store user information in PHP Session using `zend-expressive-authentication-session` is specific for this adapter. This PR enables the usage of custom `UserInterface` implementation in `zend-expressive-authentication`. I just have a look at the session implementation and maybe we can just store the `$user` instance in the session ([here](https://github.com/zendframework/zend-expressive-authentication-session/blob/master/src/PhpSession.php#L89-L92)). PHP serialize/unserialize object values automatically in the session.

weierophinney commented on src/UserRepository/Htpasswd.php:
```diff
@@ -61,10 +65,12 @@ public function authenticate(string $credential, string $password = null) : ?Use
+            $this->user->setIdentity($credential);
+            $this->user->setRoles($this->getRolesFromUser($credential));
+            return $this->user;
+        }
```
This here demonstrates why passing a factory to the repository is a better approach. If we use the same instance, we cannot have concurrent requests (ala Swoole), as the instance is in the container, and can be changed by any process running at any time.

If a factory were composed in this class instead, this would become something like:

```php
return $this->userFactory->generate($credential, $this->getRolesFromUser($credential));
```

Which would guarantee that the user generated is _unique to this request_.

geerteltink replied:
I don't think a factory is the right solution. You still need to pass the identity and roles. What about extra data, how would you pass that without knowing what data it is? And also how do you get that data again so you can persist it in a session?

weierophinney reviewed (changes requested):
I've noted a few changes, some of which we discussed elsewhere earlier today. The main points are that I think users should be immutable, and that I think we should be composing a _factory_ and not an _instance_ for the user within repositories. This allows generating the user per-request, and prevents issues of concurrency and unexpected state changes.

Additionally, I think we may need a way to store arbitrary user-related data in the `UserInterface` instances. This would be separate from the identity and roles, but could vary based on adapter. Doing so would allow us to persist that information between sessions more easily, and also ensure we can communicate that information to deeper middleware layers that may need access to it.

michalbundyra commented on src/UserRepository/Htpasswd.php:
```diff
@@ -61,10 +65,12 @@ public function authenticate(string $credential, string $password = null) : ?Use
-            && password_verify($password === null ? '' : $password, $hash)
-                ? $this->generateUser($credential)
-                : null;
+        if ($found && password_verify($password === null ? '' : $password, $hash)) {
```
We can have here `$password ?? ''` instead of `$password === null ? '' : $password`.

michalbundyra commented on src/DefaultUser.php:
```diff
@@ -0,0 +1,37 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-authentication for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (https://www.zend.com)
```
2018

michalbundyra commented on src/ConfigProvider.php:
```diff
@@ -58,11 +59,15 @@ public function getDependencies() : array
                 // AuthenticationInterface::class => Basic\BasicAccess::class,
                 // Provide an alias for the UserRepository adapter based on your application needs.
                 // UserRepositoryInterface::class => UserRepository\Htpasswd::class
+                UserInterface::class => DefaultUser::class
```
Trailing comma after the last element in the array, here and in many other places.

geerteltink commented:
> Additionally, I think we may need a way to store arbitrary user-related data in the UserInterface instances. This would be separate from the identity and roles, but could vary based on adapter. Doing so would allow us to persist that information between sessions more easily, and also ensure we can communicate that information to deeper middleware layers that may need access to it.

Have a look at this UserInterface: https://github.com/zendframework/zend-expressive-authentication/pull/25/files#diff-6e0a82375e1ea981d662cd8109eee2db

It solves a few concerns mentioned here. It make the class immutable up to a certain level and doesn't use setters. Besides that it provides a way to persist arbitrary user-related data between sessions.

ezimuel committed:
Changed DefaultUser to immutable and added UserInterfaceFactory

ezimuel committed:
Changed the doc with the latest code change

ezimuel commented:
@weierophinney I updated the PR with the following changes:
- `DefaultUser` now is a final class with immutable values;
- I provided a `UserInterfaceFactory` class that generates a `DefaultUser` instance using `$identity` and `$roles`;
- I remove the alias in the ConfigProvider;
- I updated the documentation;

@xtreamwayz I think the `UserInterface` should simply remain as it is. If you want to provide a custom implementation we can use a different `UserInterfaceFactory` implementation. Regarding the Session data, we can serialize the user's object (actually this is done automatically by PHP Session). What do you think?

weierophinney commented on docs/book/v1/intro.md:
```diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
+
+We provide a default implementation of `UserInterface` via the class `Zend\Expressive\Authentication\DefaultUser`.
+This class is a basic implementation of `UserInterface`. This class is final and
+it's immutable, that means you cannot change it state at runtime.
```
Remove the sentence "This class is a basic implementation..."; it's redundant.

Rephrase the next sentence to: "The class is final and immutable, in order to prevent runtime changes."

ezimuel replied:
Ok

weierophinney commented on docs/book/v1/intro.md:
```diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
+it's immutable, that means you cannot change it state at runtime.
+In order to set the identity and the user's role we provided a default factory
+class that generates a `UserInterface` object. This factory is
+`Zend\Expressive\Authentication\UserInterfaceFactory`. This class uses a `generate`
```
Rephrase to:

```markdown
We provide a factory class for generating `DefaultUser` instances via `Zend\Expressive\Authentication\UserFactory`.
```

Note: I think the name of the factory should either be `UserFactory` or `DefaultUserFactory`, and not contain the word `Interface`.

Additionally, if we will expect a `generate()` method, we likely need to provide an interface to describe the factory, to ensure all implementations are written the same.

ezimuel replied:
I'm rewriting the factory using `__invoke()`, so we don't need to specify another interface.

weierophinney commented on docs/book/v1/intro.md:
```diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
+
+If you want, you can customize the `UserInterfaceFactory` using your custom
+`UserInterface` implementation. You need to change the service configuration as
+follows:
```
I'd rephrase to:

```markdown
You may provide custom `UserInterface` implementations by providing your own user factory implementation. Once you have created one, change the service configuration as follows:
```

ezimuel replied:
Ok

weierophinney commented on docs/book/v1/intro.md:
````diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
+follows:
+
+```php
+// src/ConfigProvider.php
````
I would demonstrate this as straight configuration, and not as part of a `ConfigProvider`. The reasons:

- Application developers will likely want to know how to override the configuration at the application level, which is in the `config/autoload/*.php` files.
- Package developers (those creating adapters) will know that the configuration goes in a `ConfigProvider` class.

ezimuel replied:
Ok

weierophinney commented on docs/book/v1/intro.md:
```diff
@@ -75,26 +104,18 @@ Basic Access Authentication* adapter and the *htpasswd* file as the user
-
-class ConfigProvider
-{
+// src/ConfigProvider.php
```
Same comment here; do this as a dependency configuration array, not part of a `ConfigProvider`.

ezimuel replied:
Ok

weierophinney commented on src/UserInterfaceFactory.php:
```diff
@@ -0,0 +1,20 @@
+
+use Psr\Container\ContainerInterface;
+
+class UserInterfaceFactory
```
As noted previously: drop the `Interface` verbiage, and potentially add a `Default` prefix.

ezimuel replied:
Changed in `DefaultUserFactory`.

weierophinney commented on src/UserRepository/Htpasswd.php:
```diff
@@ -20,17 +21,20 @@
      * @throws Exception\InvalidConfigException
      */
-    public function __construct(string $filename)
+    public function __construct(string $filename, UserInterfaceFactory $userFactory)
```
Right here is a good example of why the factory should likely have an interface (e.g., `UserFactoryInterface`. That way, developers do not need to _extend_ the factory class, but can instead provide their own implementation of the interface.

ezimuel replied:
I changed it with a `callable`.

weierophinney reviewed (changes requested):
I've noted some additional changes.

In relation to the comments from @xtreamwayz —  should the `UserInterface` allow access to other user-related data? So, for instance:

```php
$fullName = $user->getDetail('full_name', $user->getIdentity());
$email = $user->getDetail('email', '');
$userDetails = $user->getDetails();
```

Such changes would require a third, optional argument to the user factory method to accommodate, but could answer:

- How consumers can predictably retrieve additional user details beyond the identity and roles.
- How providers can populate those details.

Thoughts?

ezimuel committed:
Changed the UserRepositoryInterface and refactored the user's factory

ezimuel commented:
@weierophinney, @xtreamwayz  I did all the proposed changes. I changed the `UserInterface` adding `getDetails()` and `getDetail(string $name, $default = null)` to be able to manage additional user information. I also added the `getUserDetails(string $identity)` function in `UserRepositoryInterface`. An implementation of this method is available in `UserRepository\PdoDatabase`, [here](https://github.com/ezimuel/zend-expressive-authentication/blob/feature/new-user-interface/src/UserRepository/PdoDatabase.php#L127-L163).

ezimuel committed:
Improved unit test coverage

geerteltink commented on src/UserRepositoryInterface.php:
```diff
@@ -26,5 +26,13 @@ public function authenticate(string $credential, string $password = null) : ?Use
      * @return string[]
      */
-    public function getRolesFromUser(string $identity) : array;
+    public function getUserRoles(string $identity) : array;
```
As asked in #29, why is this in the interface? This could be used internally in a private function in `PdoDatabase`. All the interface should care about is authenticate a user and return a UserInterface object.

ezimuel replied:
@xtreamwayz actually you right! I'm refactoring the PR removing these functions. Thanks!

geerteltink commented on src/UserRepositoryInterface.php:
```diff
@@ -26,5 +26,13 @@ public function authenticate(string $credential, string $password = null) : ?Use
+     * @param string $identity
+     * @return string[]
+     */
+    public function getUserDetails(string $identity) : array;
```
Same as above, see #29

geerteltink reviewed (commented):
It's not entirely what I had in mind but it's a nice compromise and definitely will give users the ability to use extra data.

ezimuel committed:
Added the details and changed UserRepositoryInterface

ezimuel commented:
@xtreamwayz I removed the functions from `UserRepositoryInterface`, leaving only the following method:

```php
public function authenticate(string $credential, string $password = null) : ?UserInterface;
```

Moreover, I added more unit tests for `PdoDatabase` and updated the documentation with the new interfaces.

@weierophinney I think the PR is now completed.

weierophinney commented on src/UserRepository/Htpasswd.php:
```diff
@@ -39,6 +42,7 @@ public function __construct(string $filename)
             ));
         }
         $this->filename = $filename;
+        $this->userFactory = $userFactory;
```
One thing we did with the `ResponseFactory` was to wrap this in an additional callable to provide type safety:

```php
$this->userFactory = function (string $identity, array $roles = [], array $details = []) : UserInterface use ($userFactory) {
    return $userFactory($identity, $roles, $details);
};
```

This ensures that if the callable does not return the type expected, it will raise a `TypeError`. 

I can do that change on merge.

weierophinney committed:
Provide type-safety for user factories

Wraps `$userFactory` constructor arguments in a callable that provides a
return type hint, as well as type hints and optional arguments for all
parameters.

weierophinney committed:
Rewrite section on DefaultUser and related factory

weierophinney committed:
Minor edits to user-repository doc

weierophinney committed:
Adds trailing commas

weierophinney committed:
Updates license ranges
